### ISSUE TITLE
Better support for custom reaction types

### ISSUE THREAD

**Issue by JDGrimes** [WordPoints/hooks-api#57]
While working on #12 and network mode, I entertained the possibility that we may want to introduce other custom types of reactions beyond just `standard` and `network` in the future.

I guess the closest that I got to voicing that was https://github.com/WordPoints/hooks-api/issues/12#issuecomment-163758899:

> There are definitely going to be other cases like this where we need to just get all of the reactions that affect a site, both network and standard (and maybe something else in future?).

What I had in mind is that in the future we might want to add other storage containers to a reactor (or all reactors) that would store another, special kind of reaction.

For example, say that we have a site that has trainers and trainees. We want to be able to allow the trainers to create reactions that are exclusive to their trainees. This would probably be similar to standard reactions, except that they'd have a special restriction: they'd only be active when the current user was one of the trainees of the trainer who created the reaction.

Now, we could accomplish that using a custom extension and standard reactions. Our extension would have a `should_it()` method that would check whether the reaction target was one of the trainer's trainees. The problem with that, though, is that all of the reactions created by all trainers would be displayed by the how-to-get-points shortcode. (This is actually a broader issue.)

It would also make it difficult for us to display only a particular trainer's reactions to that trainer on the page where he is supposed to manage them. We'd have a flag set in each reaction's metadata, but the only way we'd be able to filter based on that would be to pull up all of the reactions and check the value of each one. This would become expensive when there were a lot of trainers creating a lot of reactions.

A solution to that issue would be for us to keep a cache of the IDs of the reactions for each trainer, which isn't a bad solution, but it's a pain to have to implement for each "reaction type" like this.

And there will probably be other issues as well.

Also, this limits us to emulating either standard or network reactions, while it is possible we'd want something that behaves differently on multisite.

**Comment by JDGrimes** [comment]
The first step in making this a reality is making it possible to get the type of a reaction from the reaction object. Right now we can find out the reactor slug, but we don't know which storage container the reaction comes from. Because of this, the reaction ID isn't really of much use. To be able to actually identify the reaction so that we could pull it up from the reactor again, we need three items:
- The reaction ID.
- The reactor slug.
- The storage container slug.

We can get the first two from the reaction object, but not the third one. This is something that needs to be fixed for #55/#66.

Even then, we won't have the true GUID of the reaction. Since the same reaction ID can refer to different reactions in different contexts (depending on the storage method; for example, the standard storage method will contain different reactions on different sites on the network), even when we have all three of those pieces of information it only allows us to retrieve the reaction _in that particular context_. And when we start introduction custom storage types where context is determined differently, it will become impossible for us to determine what even defines that context—unless we introduce an API for that too.

So, making it possible to get a true GUID of a reaction is an important first step in making this possible. To do that we need to do two things:
- [x] Introduce a way to get the storage container slug from a reaction object.
- [x] Introduce a way to get the context ID for the storage container/reaction.

**Comment by JDGrimes** [comment]
Maybe we should just make it possible to get the storage container itself from the reaction object, then both of those pieces of information could be retrieved from the storage container. And/or we could just introduce a `get_guid()` method to the reaction class.

**Comment by JDGrimes** [comment]
I guess the current `is_network_wide()` method shared by reaction and reaction storage objects is the precursor to the introduction of a `get_context()` method. Interestingly, these methods are currently unused.

**Comment by JDGrimes** [comment]
I guess we could re-use the entity context API to supply the context information, if we wanted to. I don't think that there will be any downside to that, especially if we make it possible to override the default method.

That makes me wonder, what should happen when a reaction is out of context, or when it is retrieved in one context but the context/GUID isn't retrieved until it is in another context? Should the behavior just be undefined for now, sort of like what we do in the entity API? Or would it make more sense to set the context ID for the reaction storage object when it is constructed? But the storage objects aren't designed to be able to pull up the correct reactions no matter what context they're in. I think that right now that is overkill. So probably the best thing is just to leave that behavior as undefined for now.

I guess we could look at it this way: the reactor bootstrap currently doesn't allow a storage method to be accessed when it is out of context, although this is currently determined through the hook network mode, not the entity context API. But I think it would actually make a lot of sense to just use the entity context API here. And that way we wouldn't have this duplicate logic for determining whether we are in network context. See #61.

**Comment by JDGrimes** [comment]
In 84da9b4 I've used the entity context API, but I haven't consolidated the code from the hooks network mode to the entity context API yet.

**Comment by JDGrimes** [comment]
Network mode basically does two things, and we need to replace both of those before we can remove it:
- [x] Setting which storage group the `reactions` property refers to.
- [x] Checking whether the `standard` reactions group should be disabled.

The latter can be done through the contexts of course, and we can make it more generic, disabling direct access to any reaction group when it is out of scope.

The biggest question is how we'll determine which storage group the `reactions` property should refer to. What that property does is allow us to access the current storage group in generic reaction handling code. Which reaction group is the "current" one (currently determined by whether network-mode is on) can then be set from outside of that code.

But the question is, how do we allow the current "mode" to be set? We may actually want to start from scratch there and totally reevaluate what we're currently doing. Do we want the mode to be set across all reactions? Do we even want to still have mode switching at all?

Let's consider the reasons why we decided to keep network mode (https://github.com/WordPoints/hooks-api/issues/12#issuecomment-163414653):

> Like I said above, getting rid of network mode is going to cause us to have to write lots of duplicate code. The only way that we could avoid duplicate logic would be to abstract all kinds of things out into their own functions that pass around the reaction storage object instead of a reference to the reactor itself.

I think this will still hold true for us now.

The next question then, is whether mode-switching should be basic to the hooks API itself, as it currently is, or whether it should be a per-reactor thing. Currently network mode is global, for the following reasons (https://github.com/WordPoints/hooks-api/issues/12#issuecomment-163669249):
1. Simplicity ("we could make it more complex in the future only if it is really needed").
2. Interacting reactors ("there's the chance that we'll forget to set network mode for all reactors involved by some specific code").
3. Standard reactions cannot be loaded in the network admin anyway.

Basically, you are either in network context or you are not. So network-mode wasn't something that should be set per-reactor. That's still true for network context, of course, but it will now be determined by the entity context classes. However, just because the logic is elsewhere doesn't mean that now hook mode should be set per-reactor. I think that point no. 2 above really makes sense. Basically, I think that hook context is something that should still be set globally, not just per-reactor. The other things said about how that can be overridden and expanded later in https://github.com/WordPoints/hooks-api/issues/12#issuecomment-163669249 still apply.

The difference though is that if we allow custom reaction types to be registered optionally, when we switch to a particular mode that only exists for a single reactor, all of the other reactors will be switched too. I guess that's actually the correct behavior though; it is the way that it works now with network mode.

**Comment by JDGrimes** [comment]
But how do we automatically set the current mode? Currently we default to `standard` mode, unless we are in network context, in which case we default to `network` mode. But if we start allowing custom modes, how will we know what to default to? I guess actually that's a silly question. We should just handle the defaults the same way that we do now. When other modes are used they can be set manually.

**Comment by JDGrimes** [comment]
One more thing that I'd like to see done here before the initial release is better handling of reactor contexts. Currently reactors are assumed to be in network-wide when WordPoints is network-active, and per-site otherwise. When a reactor isn't network-wide, the `network` reaction group is disabled (since reactions won't work properly across different sites in the network). However, this is all still handled manually, outside of the entity context API. I think it would make sense for us to utilize the concept of contexts to provide handling for this. That way it will become a generic thing that can allow a reactor to be tied to any sort of context that exists, and will cause any storage groups that don't work within that context to be disabled.

How we'd do this is by adding a `$context` property to each reactor (though possibly we'll need to use a `get_context()` method instead so that the context can be determined dynamically, as it currently is—I suppose we could just set the context property in the constructor instead, but that makes the bootstrap less atomic and therefore more difficult to override).

Then I suppose we'd introduce a utility function that would be part of the entity context API that would compare to arrays of context slugs, checking whether they were equivalent, one inside the other, or each one distinct. (Maybe this would actually be several functions.)

An alternative approach would be to change the way that reaction storage groups are "registered" for each reactor. Then we could just not register a storage group for a reactor with an incompatible context. Instead of having the storage classes hard-coded as property values, we could move this information to a `Children` class registry, for example.

I guess that this wouldn't necessarily mean that we don't need to define what context a reactor exists in, though it wouldn't be as important. In most cases we'd want to know about context we could get that info from the reaction storage object, rather than the reactor. I guess the main reason that we'd want to keep this is if we still wanted to have an easy way to check if a storage group's context was compatible with the reactor's context, when registering a new storage group for a reactor. But if we're no longer registering storage groups by default when a reactor is registered, then each storage group will be registered "by hand" anyway, so it's up to the person registering them to know what context their reactor exists in.

That does lead us to the question of whether storage groups should be registered manually per-reactor, or whether there should be a global list that all reactors can have. I guess the main draw of the global list would be that it would be easier to register a storage group for every reactor. I'm not sure that would be that useful though. We'd only want to register a group for every reactor if it was going to be used by every reactor, and that would only happen if there was some kind of generic code involved, not tied to any particular reactor. That's hard for me to wrap my head around right now, since it is a totally new concept that hasn't been put to the test yet with this new API. I guess that even in a case like that though, it would be possible to register the storage group in a JIT fashion.

I think that custom reaction contexts are probably going to be useful across reactors though. Take the example of the trainer wanting to award points to his trainees. He's probably also going to want to award badges. 

In short, the modes are already a global thing, in the sense that the current mode is global itself. But we do want to be able to determine how exactly the reactions in each group are stored on a per-reactor basis. As a result, we would need to register the storage classes on a per-reactor basis. And therefore, the storage classes will have to be registered manually, and so no automated context compatibility checking would be necessary. However, we probably to still need to provide info on reactor contexts, in deference to a possible future with generic reactor handling.

**Comment by JDGrimes** [comment]
When every reaction group is registered manually, it makes it seem kind of strange to continue to offer access to each custom reaction group through the magic methods. I'm actually not sure where we would ever need to reference a particular type of reaction explicitly. Even if we did, it is now possible to switch from mode to mode without things falling apart, so we could just do that. I think until we find that this would really be a useful feature, we're better off just having the `reactions` magic method that reflects the current mode. Perhaps we could add a `get_group()` method to the reactor just in case.

**Comment by JDGrimes** [comment]
I think we've accomplished this. Anything else can be put in a new ticked.
